src/main/java/org/springframework/samples/petclinic/rest/importcsv/ImportCSV.java

```diff
@@ -26,114 +28,239 @@
+                    if(pet1.equals(pet)){
+                        clinicService.deletePet(pet1);
                         break;
                     }
                 }
             }
 
-            field = "";
-            while (i < csv.length() && (csv.charAt(i) != ';' && csv.charAt(i) != '\n')) {
-                field += csv.charAt(i++);
-            }
+            pet.setOwner(owner); // setting owner here only for the http response
+            response.add(pet);
+        }
 
-            if (pet != null) {
-                String owner = field;
-                List<Owner> matchingOwners = clinicService.findAllOwners()
-                    .stream()
-                    .filter(o -> o.getLastName().equals(owner))
-                    .collect(Collectors.toList());
-
-                if (matchingOwners.size() == 0) {
-                    HttpHeaders headers = new HttpHeaders();
-                    headers.add("errors", "Owner not found");
-                    return new ResponseEntity<List<Pet>>(headers, HttpStatus.BAD_REQUEST);
-                }
-                if (matchingOwners.size() > 1) {
-                    HttpHeaders headers = new HttpHeaders();
-                    headers.add("errors", "Owner not unique");
-                    return new ResponseEntity<List<Pet>>(headers, HttpStatus.BAD_REQUEST);
-                }
-                pet.setOwner(matchingOwners.iterator().next());
-            }
+        return new ResponseEntity<>(response, HttpStatus.OK);
+    }
 
-            if (csv.charAt(i) == ';') {
-                i++;
+    private String[] parseCommand(String command){
+        return command.split(COMMAND_DELIMITER);
+    }
 
-                field = "";
-                while (i < csv.length() && csv.charAt(i) != '\n') {
-                    field += csv.charAt(i++);
-                }
+    private Date parseDate(String dateString){
+        Date date = new Date();
+        try {
+            date = new SimpleDateFormat("yyyy-MM-dd").parse(dateString);
+        }catch (ParseException e){
+            // already handled in validateDate, should never reach this clause
+        }
 
-                if (field.toLowerCase().equals("add")) {
-                    clinicService.savePet(pet);
-                } else {
-                    for (Pet q : pet.getOwner().getPets()) {
-                        if (q.getName().equals(pet.getName())) {
-                            if (q.getType().getId().equals(pet.getType().getId())) {
-                                if (pet.getBirthDate().equals(q.getBirthDate())) {
-                                    clinicService.deletePet(q);
-                                }
-                            }
-                        }
-                    }
-                }
+        return date;
+    }
 
-            } else {
-                clinicService.savePet(pet);
-            }
-            i++;
+    private String validateCommand(String[] fields, String command){
```

Comment:
the content of this methods looks like there are some code duplications that could be avoided by creating only one, but a more general validate method that can be applied and invoked with the various fields.